File: pkg/db3/vote.go
// Copyright (c) 2022 Blockwatch Data Inc.
// Author: alex@blockwatch.cc

package db3

import (
    "blockwatch.cc/db3-near/pkg/near"
)

type Vote struct {
    AccountId near.AccountID
    ResultCID ResultCID
}

type Election struct {
    votes   []Vote
    results map[ResultCID]int
}

func NewElection() *Election {
    return &Election{
        votes:   make([]Vote, 0),
        results: make(map[ResultCID]int),
    }
}

func (e *Election) AddVote(voter near.AccountID, vote ResultCID) {
    e.votes = append(e.votes, Vote{
        AccountId: voter,
        ResultCID: vote,
    })
    e.results[vote]++
}

func (e Election) NumVoters() int {
    return len(e.votes)
}

func (e Election) NumSuperMajority() int {
    return len(e.SuperMajority())
}

func (e Election) IsUnanimous() bool {
    return len(e.results) == 1
}

func (e Election) IsSuperMajority() bool {
    numVoters := len(e.votes)
    cutoff := 200 * numVoters / 3
    for _, v := range e.results {
        if v*100 >= cutoff {
            return true
        }
    }
    return false
}

func (e Election) Minority() []Vote {
    minority := make([]Vote, 0)
    numVoters := len(e.votes)
    cutoff := 200 * numVoters / 3
    for _, vote := range e.votes {
        if e.results[vote.ResultCID]*100 < cutoff {
            minority = append(minority, vote)
        }
    }
    return minority
}

func (e Election) SuperMajority() []Vote {
    majority := make([]Vote, 0)
    numVoters := len(e.votes)
    cutoff := 200 * numVoters / 3
    for _, vote := range e.votes {
        if e.results[vote.ResultCID]*100 >= cutoff {
            majority = append(majority, vote)
        }
    }
    return majority
}

File: contract/src/vote.ts
import { UnorderedMap } from 'near-sdk-js';

export class Vote {
  account_id: string;
  result_cid: string;

  constructor({ account_id, result_cid}: { account_id: string, result_cid: string}) {
    this.account_id = account_id;
    this.result_cid = result_cid;
  }
}

export class Election {
  votes: Array<Vote>;
  results: Map<string, number>;

  constructor({ votes }:{ votes: UnorderedMap }) {
    for (let [ k, v ] of votes) {
      let account_id = k as string
      let result_cid = v as string
      this.votes.push(new Vote({ account_id, result_cid }))
      let count = this.results.get(result_cid) || 0
      this.results.set(result_cid, count + 1)
    }
  }

  numVoters(): number {
    return this.votes.length
  }

  numSuperMajority(): number {
    return this.superMajority().length
  }

  isUnanimous(): boolean {
    return this.results.size === 1
  }

  isSuperMajority(): boolean {
    let numVoters = this.votes.length
    let cutoff = 200 * numVoters / 3
    for (let [ _, v ] of this.results) {
        if (v*100 >= cutoff) {
            return true
        }
    }
    return false
  }

  minority(): Array<Vote> {
    let minority: Array<Vote>
    let numVoters = this.votes.length
    let cutoff = 200 * numVoters / 3
    for (let vote of this.votes) {
      if (this.results.get(vote.result_cid)*100 < cutoff) {
          minority.push(vote)
      }
    }
    return minority
  }

  superMajority(): Array<Vote> {
    let majority: Array<Vote>
    let numVoters = this.votes.length
    let cutoff = 200 * numVoters / 3
    for (let vote of this.votes) {
      if (this.results.get(vote.result_cid)*100 >= cutoff) {
          majority.push(vote)
      }
    }
    return majority
  }
}